src/data/history.ts

```diff
@@ -174,6 +182,134 @@ export const fetchDateWS = (
+  const params = {
+    type: "history/stream",
+    entity_ids: entityIds,
+    start_time: startTime.toISOString(),
+    end_time: endTime.toISOString(),
+    minimal_response: true,
+    no_attributes: !entityIds.some((entityId) =>
+      entityIdHistoryNeedsAttributes(hass, entityId)
+    ),
+  };
+  return hass.connection.subscribeMessage<HistoryStreamMessage>(
+    (message) => callbackFunction(message),
+    params
+  );
+};
+
+class HistoryStream {
+  hass: HomeAssistant;
+
+  hoursToShow: number;
+
+  combinedHistory: HistoryStates;
+
+  constructor(hass: HomeAssistant, hoursToShow: number) {
+    this.hass = hass;
+    this.hoursToShow = hoursToShow;
+    this.combinedHistory = {};
+  }
+
+  processMessage(streamMessage: HistoryStreamMessage): HistoryStates {
+    if (!this.combinedHistory || !Object.keys(this.combinedHistory).length) {
+      this.combinedHistory = streamMessage.states;
+      return this.combinedHistory;
+    }
+    if (!Object.keys(streamMessage.states).length) {
+      // Empty messages are still sent to
+      // indicate no more historical events
+      return this.combinedHistory;
+    }
+    const purgeBeforePythonTime =
+      (new Date().getTime() - 60 * 60 * this.hoursToShow * 1000) / 1000;
+    const newHistory: HistoryStates = {};
+    Object.keys(this.combinedHistory).forEach((entityId) => {
+      newHistory[entityId] = [];
+    });
+    Object.keys(streamMessage.states).forEach((entityId) => {
+      newHistory[entityId] = [];
+    });
+    Object.keys(newHistory).forEach((entityId) => {
+      let purgeOld = false;
+      if (
+        entityId in this.combinedHistory &&
+        entityId in streamMessage.states
+      ) {
+        newHistory[entityId] = this.combinedHistory[entityId].concat(
+          streamMessage.states[entityId]
+        );
+        if (
+          streamMessage.states[entityId][0] > this.combinedHistory[entityId][-1]
+        ) {
+          // If the history is out of order we have to sort it.
+          newHistory[entityId] = newHistory[entityId].sort(
+            (a, b) => a.lu - b.lu
+          );
+        }
+        purgeOld = true;
+      } else if (entityId in this.combinedHistory) {
+        newHistory[entityId] = this.combinedHistory[entityId];
+        purgeOld = true;
+      } else {
+        newHistory[entityId] = streamMessage.states[entityId];
+      }
+      if (purgeOld) {
+        const entityHistory = newHistory[entityId];
+        while (entityHistory[0].lu < purgeBeforePythonTime) {
+          if (entityHistory.length > 1) {
+            if (entityHistory[1].lu < purgeBeforePythonTime) {
+              newHistory[entityId].shift();
+              continue;
+            }
+          }
+          // Update the first entry to the start time state
+          // as we need to preserve the start time state and
+          // only expire the rest of the history as it ages.
+          entityHistory[0].lu = purgeBeforePythonTime;
+          break;
+        }
```

Comment:
This feels inefficient. It seems we could do something like

```ts
const lastEntry = entityHistory[entityHistory.length - 1];
const filtered = entityHistory.filter(ent => ent.lu < purgeBeforePythonTime);
if (filtered.length === 0) {
  filtered.push(lastEntry);
}
filtered[0].lu = purgeBeforePythonTime;
newHistory[entityId] = filtered;
```

This has added benefit that we don't accidentally mutate an array that is already used by a subscriber (in the case no new history, just old history).

Reply:
We can't do `ent => ent.lu < purgeBeforePythonTime` because it will remove the initial state. We need to reset the time on the first state to the start of the window so we have value at the start of the window to match the behavior of `include_start_time_state` on the backend

Reply:
Isn't that covered by storing `lastEntry` and adding it back if necessary?

Reply:
Actually it works just fine.  I read your suggestion wrong.  

Just waiting for the graph to tick forward to make sure its good to go....

Reply:
reverted for now since we need to only clear up to purge time, and than take the last item in the cleared list and make it the new start time state which makes it a bit more complex. 

Will optimize in a future PR

Reply:
https://github.com/home-assistant/frontend/pull/15153